File: shared/src/commonMain/kotlin/com/lazytravel/ui/components/layout/HeaderBar.kt
package com.lazytravel.ui.components.layout

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * HeaderBar Component - Atomic Design (Layout)
 *
 * A flexible header component with multiple variants for different screens.
 * Based on the design patterns from HTML prototypes.
 */

// ============================================================================
// Data Classes
// ============================================================================

/**
 * Represents different types of headers used throughout the app
 */
sealed class HeaderType {
    /**
     * Hero header for landing page with logo and login button
     */
    data class Hero(
        val logoText: String = "LazyTravel",
        val onLoginClick: () -> Unit = {}
    ) : HeaderType()

    /**
     * Greeting header for authenticated home screen
     */
    data class Greeting(
        val userName: String,
        val subtitle: String = "Sẵn sàng cho chuyến phiêu lưu tiếp theo?"
    ) : HeaderType()

    /**
     * Navigation header with back button and title
     */
    data class Navigation(
        val title: String,
        val subtitle: String? = null,
        val onBackClick: () -> Unit,
        val showMenu: Boolean = false,
        val onMenuClick: () -> Unit = {}
    ) : HeaderType()

    /**
     * Navigation header with trip info
     */
    data class TripNavigation(
        val tripTitle: String,
        val dateRange: String,
        val participantCount: Int,
        val onBackClick: () -> Unit
    ) : HeaderType()
}

// ============================================================================
// Main Component
// ============================================================================

/**
 * Main HeaderBar component that renders different header types
 */
@Composable
fun HeaderBar(
    type: HeaderType,
    modifier: Modifier = Modifier
) {
    when (type) {
        is HeaderType.Hero -> HeroHeaderBar(
            logoText = type.logoText,
            onLoginClick = type.onLoginClick,
            modifier = modifier
        )
        is HeaderType.Greeting -> GreetingHeaderBar(
            userName = type.userName,
            subtitle = type.subtitle,
            modifier = modifier
        )
        is HeaderType.Navigation -> NavigationHeaderBar(
            title = type.title,
            subtitle = type.subtitle,
            onBackClick = type.onBackClick,
            showMenu = type.showMenu,
            onMenuClick = type.onMenuClick,
            modifier = modifier
        )
        is HeaderType.TripNavigation -> TripNavigationHeaderBar(
            tripTitle = type.tripTitle,
            dateRange = type.dateRange,
            participantCount = type.participantCount,
            onBackClick = type.onBackClick,
            modifier = modifier
        )
    }
}

// ============================================================================
// Header Variants
// ============================================================================

/**
 * Hero header for landing page
 * Matches hero-nav from home_noauth.html
 */
@Composable
private fun HeroHeaderBar(
    logoText: String,
    onLoginClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 20.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Logo
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "✈️",
                fontSize = 24.sp
            )
            Text(
                text = logoText,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White
            )
        }

        // Login Button
        OutlinedButton(
            onClick = onLoginClick,
            colors = ButtonDefaults.outlinedButtonColors(
                containerColor = Color.White.copy(alpha = 0.2f),
                contentColor = Color.White
            ),
            border = androidx.compose.foundation.BorderStroke(1.dp, Color.White),
            contentPadding = PaddingValues(horizontal = 16.dp, vertical = 8.dp)
        ) {
            Text(
                text = "Đăng nhập",
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold
            )
        }
    }
}

/**
 * Greeting header for authenticated home
 * Matches header from home.html
 */
@Composable
private fun GreetingHeaderBar(
    userName: String,
    subtitle: String,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(Color(0xFFFAFAFA))
            .padding(horizontal = 16.dp, vertical = 16.dp)
    ) {
        Text(
            text = "Xin chào, $userName! 👋",
            fontSize = 24.sp,
            fontWeight = FontWeight.Bold,
            color = Color(0xFF212121)
        )

        Spacer(modifier = Modifier.height(4.dp))

        Text(
            text = subtitle,
            fontSize = 14.sp,
            color = Color(0xFF757575)
        )
    }
}

/**
 * Navigation header with back button
 * Matches header from create_trip.html and other pages
 */
@Composable
private fun NavigationHeaderBar(
    title: String,
    subtitle: String?,
    onBackClick: () -> Unit,
    showMenu: Boolean,
    onMenuClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(Color(0xFFFAFAFA))
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Back Button
        IconButton(
            onClick = onBackClick,
            modifier = Modifier.size(40.dp)
        ) {
            Text(
                text = "←",
                fontSize = 24.sp,
                color = Color(0xFF212121)
            )
        }

        Spacer(modifier = Modifier.width(8.dp))

        // Title & Subtitle
        Column(
            modifier = Modifier.weight(1f)
        ) {
            Text(
                text = title,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                color = Color(0xFF212121)
            )

            subtitle?.let {
                Spacer(modifier = Modifier.height(2.dp))
                Text(
                    text = it,
                    fontSize = 12.sp,
                    color = Color(0xFF757575)
                )
            }
        }

        // Optional Menu Button
        if (showMenu) {
            IconButton(
                onClick = onMenuClick,
                modifier = Modifier.size(40.dp)
            ) {
                Text(
                    text = "⋮",
                    fontSize = 24.sp,
                    color = Color(0xFF212121)
                )
            }
        }
    }
}

/**
 * Trip navigation header with trip details
 * Matches header from vote_detail.html and create_schedule.html
 */
@Composable
private fun TripNavigationHeaderBar(
    tripTitle: String,
    dateRange: String,
    participantCount: Int,
    onBackClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(Color(0xFFFAFAFA))
            .padding(horizontal = 16.dp, vertical = 12.dp)
    ) {
        // Back button and trip title row
        Row(
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(
                onClick = onBackClick,
                modifier = Modifier.size(40.dp)
            ) {
                Text(
                    text = "←",
                    fontSize = 24.sp,
                    color = Color(0xFF212121)
                )
            }

            Spacer(modifier = Modifier.width(8.dp))

            Column {
                Text(
                    text = tripTitle,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF212121)
                )

                Spacer(modifier = Modifier.height(4.dp))

                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    Text(
                        text = "📅 $dateRange",
                        fontSize = 12.sp,
                        color = Color(0xFF757575)
                    )
                    Text(
                        text = "👥 $participantCount người",
                        fontSize = 12.sp,
                        color = Color(0xFF757575)
                    )
                }
            }
        }
    }
}

// ============================================================================
// Preview/Usage Examples (for documentation)
// ============================================================================

/**
 * Usage Examples:
 *
 * // Hero header for landing page
 * HeaderBar(
 *     type = HeaderType.Hero(
 *         logoText = "LazyTravel",
 *         onLoginClick = { /* navigate to login */ }
 *     )
 * )
 *
 * // Greeting header for home
 * HeaderBar(
 *     type = HeaderType.Greeting(
 *         userName = "Minh",
 *         subtitle = "Sẵn sàng cho chuyến phiêu lưu tiếp theo?"
 *     )
 * )
 *
 * // Navigation header with back button
 * HeaderBar(
 *     type = HeaderType.Navigation(
 *         title = "Tạo Chuyến Đi Mới",
 *         subtitle = "Lên kế hoạch chi tiết cho chuyến đi",
 *         onBackClick = { /* navigate back */ }
 *     )
 * )
 *
 * // Trip navigation header
 * HeaderBar(
 *     type = HeaderType.TripNavigation(
 *         tripTitle = "Chuyến đi Nha Trang",
 *         dateRange = "15-18/12/2024",
 *         participantCount = 5,
 *         onBackClick = { /* navigate back */ }
 *     )
 * )
 */

File: iosApp/iosApp/Components/Organisms/HeaderBar.swift
import SwiftUI

// MARK: - Header Types
enum HeaderType {
    case hero(logoText: String, onLoginTap: () -> Void)
    case greeting(userName: String, subtitle: String)
    case navigation(title: String, subtitle: String?, onBackTap: () -> Void, showMenu: Bool, onMenuTap: () -> Void)
    case tripNavigation(tripTitle: String, dateRange: String, participantCount: Int, onBackTap: () -> Void)
}

// MARK: - HeaderBar Component
struct HeaderBar: View {
    let type: HeaderType

    var body: some View {
        switch type {
        case .hero(let logoText, let onLoginTap):
            HeroHeader(logoText: logoText, onLoginTap: onLoginTap)
        case .greeting(let userName, let subtitle):
            GreetingHeader(userName: userName, subtitle: subtitle)
        case .navigation(let title, let subtitle, let onBackTap, let showMenu, let onMenuTap):
            NavigationHeader(title: title, subtitle: subtitle, onBackTap: onBackTap, showMenu: showMenu, onMenuTap: onMenuTap)
        case .tripNavigation(let tripTitle, let dateRange, let participantCount, let onBackTap):
            TripNavigationHeader(tripTitle: tripTitle, dateRange: dateRange, participantCount: participantCount, onBackTap: onBackTap)
        }
    }
}

// MARK: - Hero Header (Landing Page)
private struct HeroHeader: View {
    let logoText: String
    let onLoginTap: () -> Void

    var body: some View {
        HStack {
            // Logo
            HStack(spacing: 8) {
                Text("✈️")
                    .font(.system(size: 24))
                Text(logoText)
                    .font(.system(size: 18, weight: .bold))
                    .foregroundColor(.white)
            }

            Spacer()

            // Login Button
            Button(action: onLoginTap) {
                Text("Đăng nhập")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 8)
                    .background(Color.white.opacity(0.2))
                    .cornerRadius(8)
                    .overlay(
                        RoundedRectangle(cornerRadius: 8)
                            .stroke(Color.white, lineWidth: 1)
                    )
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 20)
        .background(
            LinearGradient(
                colors: [Color(hex: "FF6B35"), Color(hex: "F7931E")],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
        )
    }
}

// MARK: - Greeting Header (Authenticated Home)
private struct GreetingHeader: View {
    let userName: String
    let subtitle: String

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("Xin chào, \(userName)! 👋")
                .font(.system(size: 24, weight: .bold))
                .foregroundColor(Color(hex: "212121"))

            Text(subtitle)
                .font(.system(size: 14))
                .foregroundColor(Color(hex: "757575"))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 16)
        .padding(.vertical, 16)
        .background(Color(hex: "FAFAFA"))
    }
}

// MARK: - Navigation Header (Standard Pages)
private struct NavigationHeader: View {
    let title: String
    let subtitle: String?
    let onBackTap: () -> Void
    let showMenu: Bool
    let onMenuTap: () -> Void

    var body: some View {
        HStack(spacing: 8) {
            // Back Button
            Button(action: onBackTap) {
                Text("←")
                    .font(.system(size: 24))
                    .foregroundColor(Color(hex: "212121"))
                    .frame(width: 40, height: 40)
            }

            // Title & Subtitle
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.system(size: 18, weight: .bold))
                    .foregroundColor(Color(hex: "212121"))

                if let subtitle = subtitle {
                    Text(subtitle)
                        .font(.system(size: 12))
                        .foregroundColor(Color(hex: "757575"))
                }
            }

            Spacer()

            // Optional Menu Button
            if showMenu {
                Button(action: onMenuTap) {
                    Text("⋮")
                        .font(.system(size: 24))
                        .foregroundColor(Color(hex: "212121"))
                        .frame(width: 40, height: 40)
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
        .background(Color(hex: "FAFAFA"))
    }
}

// MARK: - Trip Navigation Header
private struct TripNavigationHeader: View {
    let tripTitle: String
    let dateRange: String
    let participantCount: Int
    let onBackTap: () -> Void

    var body: some View {
        HStack(spacing: 8) {
            // Back Button
            Button(action: onBackTap) {
                Text("←")
                    .font(.system(size: 24))
                    .foregroundColor(Color(hex: "212121"))
                    .frame(width: 40, height: 40)
            }

            // Trip Info
            VStack(alignment: .leading, spacing: 4) {
                Text(tripTitle)
                    .font(.system(size: 16, weight: .bold))
                    .foregroundColor(Color(hex: "212121"))

                HStack(spacing: 12) {
                    Text("📅 \(dateRange)")
                        .font(.system(size: 12))
                        .foregroundColor(Color(hex: "757575"))

                    Text("👥 \(participantCount) người")
                        .font(.system(size: 12))
                        .foregroundColor(Color(hex: "757575"))
                }
            }

            Spacer()
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
        .background(Color(hex: "FAFAFA"))
    }
}

// MARK: - Preview
struct HeaderBar_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 20) {
            HeaderBar(type: .greeting(userName: "Minh", subtitle: "Sẵn sàng cho chuyến phiêu lưu tiếp theo?"))

            HeaderBar(type: .navigation(title: "Tạo Chuyến Đi Mới", subtitle: "Lên kế hoạch chi tiết", onBackTap: {}, showMenu: true, onMenuTap: {}))

            HeaderBar(type: .tripNavigation(tripTitle: "Chuyến đi Nha Trang", dateRange: "15-18/12/2024", participantCount: 5, onBackTap: {}))

            Spacer()
        }
    }
}
